test-libz-rs-sys/src/gz.rs

```diff
@@ -6,42 +6,83 @@ use std::ffi::CString;
 use std::ptr;
 
 // Generate a file path relative to the project's root
-macro_rules! path {
+macro_rules! crate_path {
     ($str:expr) => {
         concat!(env!("CARGO_MANIFEST_DIR"), "/", $str)
     };
 }
 
-unsafe fn test_open(path: &str, mode: &str, should_succeed: bool) {
-    let cpath = CString::new(path).unwrap();
-    let cmode = CString::new(mode).unwrap();
-    let handle = gzopen(cpath.as_ptr(), cmode.as_ptr());
-    assert_eq!(should_succeed, !handle.is_null(), "{path} {mode}");
-    if !handle.is_null() {
-        assert_eq!(gzclose(handle), Z_OK);
-    }
+// Generate a file path relative to a specified directory prefix
+macro_rules! path {
+    ($prefix:expr, $str:expr) => {{
+        let mut path_buf = $prefix.to_path_buf();
+        path_buf.push($str);
+        path_buf.as_path().to_str().unwrap().to_owned()
+    }};
+}
+
+macro_rules! test_open {
+    ($path:expr, $mode:expr, $should_succeed:expr) => {
+        let cpath = CString::new($path).unwrap();
+        let cmode = CString::new($mode).unwrap();
+        let handle = unsafe { gzopen(cpath.as_ptr(), cmode.as_ptr()) };
+        assert_eq!($should_succeed, !handle.is_null(), "{} {}", $path, $mode);
+        if !handle.is_null() {
+            assert_eq!(unsafe { gzclose(handle) }, Z_OK);
+        }
+    };
 }
 
 #[test]
 fn open_close() {
-    unsafe {
-        // Open a valid file for reading
-        test_open(path!("src/test-data/issue-109.gz"), "r", true);
+    // Open a valid file for reading
+    test_open!(crate_path!("src/test-data/issue-109.gz"), "r", true);
+
+    // "b" for binary mode is optional
+    test_open!(crate_path!("src/test-data/issue-109.gz"), "rb", true);
+
+    // Mode must include r, w, or a
+    test_open!(crate_path!("src/test-data/issue-109.gz"), "", false);
+    test_open!(crate_path!("src/test-data/issue-109.gz"), "e", false);
+
+    // For zlib-ng compatibility, mode can't specify transparent read
+    test_open!(crate_path!("src/test-data/issue-109.gz"), "Tr", false);
+
+    // Read of a nonexistent file should fail
+    test_open!(crate_path!("src/test-data/no-such-file.gz"), "r", false);
+
+    // Closing a null file handle should return an error instead of crashing
+    assert_eq!(unsafe { gzclose(ptr::null_mut()) }, Z_STREAM_ERROR);
+}
+
+#[test]
+fn create() {
+    // Create a temporary directory that will be automatically removed when
+    // temp_dir goes out of scope.
+    let temp_dir_path = if cfg!(target_os = "wasi") {
+        std::path::PathBuf::from("/tmp/")
+    } else {
+        std::env::temp_dir()
+    };
+
+    let temp_dir = tempfile::TempDir::new_in(temp_dir_path).unwrap();
+    let temp_path = temp_dir.path();
```

Comment:
so yeah we'll have to use `TempDir::new_in` instead of `TempDir::new` throughout. We can probably extract this logic into a function when we add more tests.